File: 3-leetcode/02-medium/dart/1140-stone-game-ii.dart
class Solution {
  Map<String, int> memo = {};
  int stoneGameII(List<int> piles) {
    int n = piles.length;
    List<int> suffixSum = List.filled(n, 0);
    suffixSum[n - 1] = piles[n - 1];
    for (int i = n - 2; i >= 0; i--) {
      suffixSum[i] = piles[i] + suffixSum[i + 1];
    }
    int dp(int i, int m) {
      if (i >= n) {
        return 0;
      }
      if (memo.containsKey('$i-$m')) {
        return memo['$i-$m']!;
      }
      int maxStones = 0;
      for (int x = 1; x <= 2 * m && i + x <= n; x++) {
        maxStones = max(maxStones, suffixSum[i] - dp(i + x, max(m, x)));
      }
      memo['$i-$m'] = maxStones;
      return maxStones;
    }

    return dp(0, 1);
  }
}

File: 3-leetcode/02-medium/js/1140-stone-game-ii.ts
const stoneGameII = (piles: number[]) => {
  const memo = new Map<string, number>();
  let n = piles.length;
  const suffixSum = Array.from({ length: n }, () => 0);
  suffixSum[n - 1] = piles[n - 1];
  for (let i = n - 2; i >= 0; i--) {
    suffixSum[i] = piles[i] + suffixSum[i + 1];
  }
  const dp = (i: number, m: number) => {
    if (i >= n) {
      return 0;
    }

    if (memo.has(`${i}-${m}`)) {
      return memo.get(`${i}-${m}`);
    }

    let maxStones = 0;

    for (let x = 1; x <= 2 * m && i + x <= n; x++) {
      maxStones = Math.max(maxStones, suffixSum[i] - dp(i + x, Math.max(m, x)));
    }

    memo.set(`${i}-${m}`, maxStones);

    return maxStones;
  };

  return dp(0, 1);
};
